### PR TITLE
Optimize for bool attributes. 

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp [hunk 1]
```diff
@@ -468,7 +488,10 @@ AttributeBlueprint::setup(const fef::IIndexEnvironment & env,
                    "the value at the given index of an array attribute, "
                    "the given key of a weighted set attribute, or"
                    "the tensor of a tensor attribute", output_type);
-    if (!_tensorType.is_tensor()) {
+    const fef::FieldInfo * fInfo = env.getFieldByName(_attrName);
+    if (!_tensorType.is_tensor() &&
+        !((fInfo->collection() == schema::CollectionType::SINGLE) && (fInfo->get_data_type() == schema::DataType::BOOL)))
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/attributefeature.cpp]: please simplify this condition to make it clearer what we are checking

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp [hunk 2]
```diff
@@ -365,6 +379,12 @@ createAttributeExecutor(const IAttributeVector *attribute, const vespalib::strin
             { SingleValueExecutorCreator<IntegerAttributeTemplate<int8_t>> creator;       if (creator.handle(attribute)) return creator.create(stash); }
             { SingleValueExecutorCreator<IntegerAttributeTemplate<int32_t>> creator;      if (creator.handle(attribute)) return creator.create(stash); }
             { SingleValueExecutorCreator<IntegerAttributeTemplate<int64_t>> creator;      if (creator.handle(attribute)) return creator.create(stash); }
+            {
+                auto boolAttribute = dynamic_cast<const SingleBoolAttribute *>(attribute);
+                if (boolAttribute) {
+                    return stash.create<BoolAttributeExecutor>(*boolAttribute);
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/features/attributefeature.cpp]: This should be possible if and only if we also chose to drop additional outputs.
It feels a bit fragile to check meta-data in one end and actual attribute implementation class in the other end.

**Review by havardpe — COMMENTED** [review]
We need to separate executors with different number of outputs better. Right now the fallback for not being able to find the flag attribute is to create an executor producing 4 zero values (which would overwrite outside our 1 allocated output)

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/attributefeature.cpp]: Done

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp (hunk 2 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/attributefeature.cpp]: Fixed and added some more sanity checks.

**Comment by baldersheim** [comment]
@havardpe Now you can have a look again.

**Review by havardpe — COMMENTED** [review]
still need fever zero fallback values for bool attributes

File: searchlib/src/vespa/searchlib/features/attributefeature.cpp [hunk 3]
```diff
@@ -337,15 +351,17 @@ struct MultiValueExecutorCreator {
 };
 
 fef::FeatureExecutor &
-createAttributeExecutor(const IAttributeVector *attribute, const vespalib::string &attrName, const vespalib::string &extraParam, vespalib::Stash &stash)
+createAttributeExecutor(uint32_t numOutputs, const IAttributeVector *attribute, const vespalib::string &attrName, const vespalib::string &extraParam, vespalib::Stash &stash)
 {
     if (attribute == nullptr) {
         LOG(warning, "The attribute vector '%s' was not found in the attribute manager, returning default values.",
                 attrName.c_str());
         std::vector<feature_t> values(4, 0.0f);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/features/attributefeature.cpp]: Correct, done

**Comment by baldersheim** [comment]
Superceeded by rebased variant #12464

**Pull request closed without merging**